### PR TITLE
Updating marks through API

### PR DESCRIPTION
**Pull request by hootiehoo**
Allows TAs/Admins to update marks via api.
 
Valid requests are of the form:
curl -H "Authorization: MarkUsAuth auth_key" -X PUT "http://example.com/api/assignments/assignment_id/groups/group_id/update_marks?criterion1=4&criterion2=3"
where criterionX is a properly formatted title of the desired criterion. Eg if the criterion is "Aspernatur dolorem.", we would put ?Aspernatur%20dolorem.=4 .

### CODE REVIEW

File: app/controllers/api/groups_controller.rb [hunk 1]
```diff
@@ -80,5 +80,54 @@ def include_students(fields)
       end
     end
 
+    # Update the group's marks for the given assignment.
+    def update_marks
+      assignment = Assignment.find_by_id(params[:assignment_id])
```
Comment (david-yz-liu) [inline app/controllers/api/groups_controller.rb]: Just use `find`

File: app/controllers/api/groups_controller.rb [hunk 2]
```diff
@@ -80,5 +80,54 @@ def include_students(fields)
...
+          'No group exists with that id' }, status: 404
+        return
+      elsif group.submissions.first.nil?
+        render 'shared/http_status', locals: { code: '404', message:
+          'No submissions exist for that group' }, status: 404
+        return
+      end
+
+      all_criteria = assignment.rubric_criteria
+      all_criteria.push(*assignment.flexible_criteria)
```
Comment (david-yz-liu) [inline app/controllers/api/groups_controller.rb]: I think you want `concat`, not `push`

File: app/controllers/api/groups_controller.rb [hunk 3]
```diff
@@ -80,5 +80,54 @@ def include_students(fields)
...
+          'No assignment exists with that id' }, status: 404
+        return
+      end
+
+      group = Group.find_by_id(params[:id])
+      if group.nil?
+        render 'shared/http_status', locals: { code: '404', message:
+          'No group exists with that id' }, status: 404
+        return
+      elsif group.submissions.first.nil?
```
Comment (david-yz-liu) [inline app/controllers/api/groups_controller.rb]: We really want to be using the `grouping` properties `has_submission?` and `current_submission_used` here. This code may be incorrect when there are multiple submissions.

File: app/controllers/api/groups_controller.rb [hunk 4]
```diff
@@ -80,5 +80,54 @@ def include_students(fields)
...
+      all_criteria = assignment.rubric_criteria
+      all_criteria.push(*assignment.flexible_criteria)
+      if all_criteria.empty?
+        render 'shared/http_status', locals: { code: '404', message:
+          'No criteria exist for that assignment' }, status: 404
+        return
+      end
+
+      result = group.submissions.first.get_latest_result
+      params.keys.each do |param_key|
```
Comment (david-yz-liu) [inline app/controllers/api/groups_controller.rb]: This loop structure is unnecessary.

Set up a query like `RubricCriterion.where(assignment: assignment, rubric_criterion_name: params[:name])` to do the search. (Need another one for FlexibleCriterion too.)

File: app/controllers/api/groups_controller.rb [hunk 5]
```diff
@@ -80,5 +80,62 @@ def include_students(fields)
...
+      end
+
+      group = Group.find(params[:id])
+      if group.nil?
+        render 'shared/http_status', locals: { code: '404', message:
+          'No group exists with that id' }, status: 404
+        return
+      end
+      if group.groupings.first.has_submission?
+        result = group.groupings
```
Comment (david-yz-liu) [inline app/controllers/api/groups_controller.rb]: Use `group.get_grouping_for`

File: app/controllers/api/groups_controller.rb [hunk 6]
```diff
@@ -80,5 +80,62 @@ def include_students(fields)
...
+        mark_to_change = result.marks
+                               .where(markable_id: crit.id)
+                               .first
+        if crit.is_a?(FlexibleCriterion)
+          mark_to_change.mark = params[crit.flexible_criterion_name].to_f
+        else
+          mark_to_change.mark = params[crit.rubric_criterion_name].to_i
+        end
+        unless mark_to_change.save
+          # Some error occurred
```
Comment (david-yz-liu) [inline app/controllers/api/groups_controller.rb]: Don't put this check in.

**Comment by david-yz-liu** [comment]
@hootiehoo the function is okay, but one improvement would be taking the actual mark updating part (*after* the database has been queried) and putting it into a separate function.

File: app/controllers/api/groups_controller.rb [hunk 7]
```diff
@@ -80,5 +80,59 @@ def include_students(fields)
...
+                      .current_submission_used
+                      .get_latest_result
+      else
+        render 'shared/http_status', locals: { code: '404', message:
+          'No submissions exist for that group' }, status: 404
+        return
+      end
+
+      matched_criteria = RubricCriterion
+                           .where(rubric_criterion_name: params.keys)
```
Comment (david-yz-liu) [inline app/controllers/api/groups_controller.rb]: I think I meant to make a comment here and then didn't. This isn't searching criteria by a particular assignment, which isn't good if two assignments have a criterion with the same name.

File: app/controllers/api/groups_controller.rb [hunk 8]
```diff
@@ -73,6 +73,15 @@ def show
       end
     end
 
+
+    def get_group_by_name
+    end
+
+
```
Comment (hootiehoo) [inline app/controllers/api/groups_controller.rb]: Oops, these shouldn't be here.

**Comment by david-yz-liu** [comment]
Thanks, @hootiehoo!

**Pull request merged**